### ISSUE TITLE
[JENKINS-40164] Obtain parallel warnings results in pipeline

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/warnings-ng-plugin#2579]
<p>I currently use the warnings plugin in a pipeline build by simply calling the build step. Doing this sets the build result to unstable, so I was checking that during the pipeline to determine whether some extra steps to run.</p>

<p>Specifically, I have a parallel build, which builds on multiple different configurations in parallel, and I want to rename the log file for each build to "unstable" only if that configuration got warnings. However, the build result is global for the whole pipeline. How can I obtain the number of warnings or the "result" of the warnings step in such a way that each parallel invocation can get its own results? This allows me to rename my logged files so that users who look at the final results can quickly determine which configurations were faulty?</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jekeller.png" title="jekeller's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jekeller">jekeller</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164" target="_blank">Obtain parallel warnings results in pipeline</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=drulli">drulli</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>component(s)</b>: analysis-core-plugin, warnings-plugin
<li><b>label(s)</b>: analysis-core-2.0
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-05-19T12:36:05+00:00
<li><b>votes</b>: 4
<li><b>watchers</b>: 8
<li><b>imported</b>: 2025-12-09
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I currently use the warnings plugin in a pipeline build by simply calling the build step. Doing this sets the build result to unstable, so I was checking that during the pipeline to determine whether some extra steps to run.</p>

<p>Specifically, I have a parallel build, which builds on multiple different configurations in parallel, and I want to rename the log file for each build to "unstable" only if that configuration got warnings. However, the build result is global for the whole pipeline. How can I obtain the number of warnings or the "result" of the warnings step in such a way that each parallel invocation can get its own results? This allows me to rename my logged files so that users who look at the final results can quickly determine which configurations were faulty?</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [Page](https://wiki.jenkins.io/pages/viewpage.action?pageId=133956397)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> drulli:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=281865&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-281865">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Can you please describe in more detail what should work? Maybe you can write a small sketch of the scripts you would expect to work. I do not use pipelines on my own so it is hard to understand the actual requirement. </p>
</pre>
</details>

</ul>
</details>

<p>Can you please describe in more detail what should work? Maybe you can write a small sketch of the scripts you would expect to work. I do not use pipelines on my own so it is hard to understand the actual requirement. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> drulli:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=281919&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-281919">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Please also check if this is a duplicate of the referenced issues.</p>
</pre>
</details>

</ul>
</details>

<p>Please also check if this is a duplicate of the referenced issues.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jekeller.png" title="jekeller's avatar" /> jekeller:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=281946&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-281946">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p> <a href="https://issues.jenkins.io/browse/JENKINS-30551" title="Make Warnings Plugin compatible with parallel Workflow" class="issue-link" data-issue-key="JENKINS-30551"><del>JENKINS-30551</del></a> is related but not identical to this issue.</p>

<p>Basically, what I want to be able to do is to run the warnings plugin inside a parallel branch, so that I can scan warnings for a particular log file generated differently for each parallel branch. In this case, each branch is running on a separate node, and each node is a different version of the Linux  Kernel, I am compiling a driver.</p>

<p>I need to be able to determine the outcome of the warnings step inside the parallel branch. Unfortunately, the warnings plugin sets the entire job status, so if I just check the buildStatus variable directly, I am unable to determine whether the failure was this node or some other node, as the process is being run in parallel.</p>

<p>Ideally, a separate pipeline specific step would be introduced which can run the warnings plugin and throws an exception when it fails. I could then catch the exception in the pipeline script, and handle it in various ways. My current job wants to rename a file to indicate whether the log has warnings or not.</p>

<p>Finally,  <a href="https://issues.jenkins.io/browse/JENKINS-30551" title="Make Warnings Plugin compatible with parallel Workflow" class="issue-link" data-issue-key="JENKINS-30551"><del>JENKINS-30551</del></a> is about making the display of multiple parallel branches work correctly which is a good thing but is not exactly related to my problem.</p>
</pre>
</details>

</ul>
</details>

<p> <a href="https://issue-redirect.jenkins.io/issue/30551" title="Make Warnings Plugin compatible with parallel Workflow" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/30551">JENKINS-30551</a>"><del>JENKINS-30551</del></a> is related but not identical to this issue.</p>

<p>Basically, what I want to be able to do is to run the warnings plugin inside a parallel branch, so that I can scan warnings for a particular log file generated differently for each parallel branch. In this case, each branch is running on a separate node, and each node is a different version of the Linux  Kernel, I am compiling a driver.</p>

<p>I need to be able to determine the outcome of the warnings step inside the parallel branch. Unfortunately, the warnings plugin sets the entire job status, so if I just check the buildStatus variable directly, I am unable to determine whether the failure was this node or some other node, as the process is being run in parallel.</p>

<p>Ideally, a separate pipeline specific step would be introduced which can run the warnings plugin and throws an exception when it fails. I could then catch the exception in the pipeline script, and handle it in various ways. My current job wants to rename a file to indicate whether the log has warnings or not.</p>

<p>Finally,  <a href="https://issue-redirect.jenkins.io/issue/30551" title="Make Warnings Plugin compatible with parallel Workflow" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/30551">JENKINS-30551</a>"><del>JENKINS-30551</del></a> is about making the display of multiple parallel branches work correctly which is a good thing but is not exactly related to my problem.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> drulli:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=282054&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-282054">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I see. This is not supported yet. (Is there another plug-in that works this way?)</p>
</pre>
</details>

</ul>
</details>

<p>I see. This is not supported yet. (Is there another plug-in that works this way?)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jekeller.png" title="jekeller's avatar" /> jekeller:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=282070&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-282070">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>For example, in a pipeline, you can do something like:</p>

<p>try {
  sh "some-command"
catch (err) {
  "do stuff"
}</p>

<p>In a regular build, you would run shell script as part of the main build steps, and when the shell returns with a negative error code, the build result would become failure.</p>

<p>However, the warnings plugin is run via the generic "step()" pipeline, and instead of returning an error or throwing an exception, the step literally sets the build status. This doesn't work in parallel since the whole build has a single status, and each parallel branch might result in different build status, and thus conflict.</p>

<p>What I want is a new "scanWarnings" step which is a pipeline specific step we could call, and it would throw an exception, which we could catch in the pipeline code.</p>

<p>It might be worth getting some input from someone who uses pipelines more?</p>
</pre>
</details>

</ul>
</details>

<p>For example, in a pipeline, you can do something like:</p>

<p>try {<br/>
  sh "some-command"<br/>
catch (err) {<br/>
  "do stuff"<br/>
}</p>

<p>In a regular build, you would run shell script as part of the main build steps, and when the shell returns with a negative error code, the build result would become failure.</p>

<p>However, the warnings plugin is run via the generic "step()" pipeline, and instead of returning an error or throwing an exception, the step literally sets the build status. This doesn't work in parallel since the whole build has a single status, and each parallel branch might result in different build status, and thus conflict.</p>

<p>What I want is a new "scanWarnings" step which is a pipeline specific step we could call, and it would throw an exception, which we could catch in the pipeline code.</p>

<p>It might be worth getting some input from someone who uses pipelines more?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> drulli:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=282152&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-282152">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I found some documentation about doing this: <a href="https://github.com/jenkinsci/workflow-step-api-plugin/blob/master/README.md" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/workflow-step-api-plugin/blob/master/README.md</a>.</p>
</pre>
</details>

</ul>
</details>

<p>I found some documentation about doing this: <a href="https://github.com/jenkinsci/workflow-step-api-plugin/blob/master/README.md" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/workflow-step-api-plugin/blob/master/README.md</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jekeller.png" title="jekeller's avatar" /> jekeller:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=300924&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-300924">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Right, that outline is how to add a new step which is pretty much what we want to do, I think. Ideally we would have some way of actually obtaining the results of the warnings directly in groovy so that the groovy code could make all the decisions based on those results.</p>
</pre>
</details>

</ul>
</details>

<p>Right, that outline is how to add a new step which is pretty much what we want to do, I think. Ideally we would have some way of actually obtaining the results of the warnings directly in groovy so that the groovy code could make all the decisions based on those results.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jekeller.png" title="jekeller's avatar" /> jekeller:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=307077&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307077">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Specifically, what I want is to be able to get the WarningsResult for a given run of the warnings step, so I can encode the logic of how to respond to warnings inside the pipeline code.</p>

<p>I need to be able to run 8 nodes in parallel, each compiling on a different distribution, and see warnings for each one, and I'd like to be able to individually label the files as "UNSTABLE" without racing against other warning results. The current problem is that currentBuild.result doesn't work well because any of the 8 nodes could finish first, and make the build result Unstable, which means I could label the wrong files as unstable, when that node didn't produce warnings.</p>

<p> </p>

<p>I want a way to programatically say "this node had warnings, and this one didn't"</p>
</pre>
</details>

</ul>
</details>

<p>Specifically, what I want is to be able to get the WarningsResult for a given run of the warnings step, so I can encode the logic of how to respond to warnings inside the pipeline code.</p>

<p>I need to be able to run 8 nodes in parallel, each compiling on a different distribution, and see warnings for each one, and I'd like to be able to individually label the files as "UNSTABLE" without racing against other warning results. The current problem is that currentBuild.result doesn't work well because any of the 8 nodes could finish first, and make the build result Unstable, which means I could label the wrong files as unstable, when that node didn't produce warnings.</p>

<p> </p>

<p>I want a way to programatically say "this node had warnings, and this one didn't"</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> drulli:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=307079&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307079">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This is quite related to the other pipeline requirements. I think I need to split the whole reporting code into several steps that can be used independently from each other in a pipeline. </p>
</pre>
</details>

</ul>
</details>

<p>This is quite related to the other pipeline requirements. I think I need to split the whole reporting code into several steps that can be used independently from each other in a pipeline. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jekeller.png" title="jekeller's avatar" /> jekeller:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=307080&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307080">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Note that this can't even be done currently with the currentBuild.rawBuild manually. Even if you manage to get the actions, there's no way to tell which WarningsResultAction actually contains the warnings you care about, since the ParserResult doesn't store which workspace files it scans.</p>
</pre>
</details>

</ul>
</details>

<p>Note that this can't even be done currently with the currentBuild.rawBuild manually. Even if you manage to get the actions, there's no way to tell which WarningsResultAction actually contains the warnings you care about, since the ParserResult doesn't store which workspace files it scans.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jekeller.png" title="jekeller's avatar" /> jekeller:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=307082&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307082">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Actually, I think all you need to do is create a step which takes a parser configuration and creatse a FileParser object, execute it, and return the ParserResult... That might not be too complicated.</p>
</pre>
</details>

</ul>
</details>

<p>Actually, I think all you need to do is create a step which takes a parser configuration and creatse a FileParser object, execute it, and return the ParserResult... That might not be too complicated.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> drulli:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=307092&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307092">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>That is exactly what I mean. I think I will write down these ideas in a wiki page so that everybody can help to refine these requirements before I can start with an implementation...</p>
</pre>
</details>

</ul>
</details>

<p>That is exactly what I mean. I think I will write down these ideas in a wiki page so that everybody can help to refine these requirements before I can start with an implementation...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> drulli:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=307149&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307149">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm trying to consolidate the requirements for the static analysis suite in pipeline jobs in a <a href="https://wiki.jenkins.io/display/JENKINS/Static+Analysis+in+Pipelines" class="external-link" target="_blank" rel="nofollow noopener">wiki page</a>. Can you please read it carefully and comment or change it accordingly? </p>
</pre>
</details>

</ul>
</details>

<p>I'm trying to consolidate the requirements for the static analysis suite in pipeline jobs in a <a href="https://wiki.jenkins.io/display/JENKINS/Static+Analysis+in+Pipelines" class="external-link" target="_blank" rel="nofollow noopener">wiki page</a>. Can you please read it carefully and comment or change it accordingly? </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/drulli.png" title="drulli's avatar" /> drulli:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-40164?focusedId=349102&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-349102">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Released in 5.0.0-beta2.</p>
</pre>
</details>

</ul>
</details>

<p>Released in 5.0.0-beta2.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/30551">JENKINS-30551</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/37325">JENKINS-37325</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `depends_on` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/49938">JENKINS-49938</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `epic` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/49911">JENKINS-49911</a>]</i>
